### PR TITLE
Pretty: Expr.Do

### PR DESCRIPTION
**Pull request by HoshinoTented**
My idea is that render an `Expr.Do` as Single Line or Multi-Line (see below)

Single Line (all bindings in one line):

```
do { x <- xs, y <- ys, return (x + y) }
```

Multi-Line (one binding each line):

```
do {
  x <- xs,
  y <- ys,
  return (x + y)
}
```

I dislike (simply `Doc.sep` when `lineRemaining` is insufficient):
```
do { x <- xs, y <- ys,
  return (x + y)
}
```

Can we advise the PrettyPrinter to flat something even though the page width is enough?

### CODE REVIEW

File: base/src/main/java/org/aya/distill/ConcreteDistiller.java [hunk 1]
```diff
@@ -153,7 +154,23 @@ yield visitCalls(false, fn, (nc, l) -> l.toDoc(options), outer,
...
+          } else {
+            return visitDoBind(bind);
+          }
+        });
+
+        // Either not flat (single line) or full flat
+        yield Doc.flatAltBracedBlock(
+          Doc.commaList(doBlockDoc),
+          Doc.vcommaList(
+            doBlockDoc.map(x -> Doc.nest(2, x))     // TODO[hoshino]: constant indent?
```
Comment (ice1000) [inline base/src/main/java/org/aya/distill/ConcreteDistiller.java:170]: I think this is good for now

File: pretty/src/main/java/org/aya/pretty/doc/Doc.java [hunk 2]
```diff
@@ -242,6 +242,18 @@ record PageWidth(@NotNull IntFunction<Doc> docBuilder) implements Doc {
...
     return wrap("{", "}", doc);
   }
+
+  /**
+   * Either `{ defaultDoc }` or `{\nflatDoc\n}`
+   * TODO: There should be a better impl for this! For example, a new Doc element class
+   */
+  static @NotNull Doc flatAltBracedBlock(Doc defaultDoc, Doc flatDoc) {
+    return flatAlt(
+      stickySep(Doc.plain("{"), defaultDoc, Doc.plain("}")),
```
Comment (ice1000) [inline pretty/src/main/java/org/aya/pretty/doc/Doc.java:252]: Use `Doc.symbol("{")`

File: pretty/src/main/java/org/aya/pretty/doc/Doc.java [hunk 3]
```diff
@@ -518,6 +530,10 @@ record PageWidth(@NotNull IntFunction<Doc> docBuilder) implements Doc {
     return join(COMMA, docs);
   }
 
+  @Contract("_ -> new") static @NotNull Doc vcommaList(@NotNull SeqLike<Doc> docs) {
+    return join(cat(plain(","), line()), docs);
```
Comment (ice1000) [inline pretty/src/main/java/org/aya/pretty/doc/Doc.java:534]: `plain(",")` -> `COMMA`

File: pretty/src/main/java/org/aya/pretty/doc/Doc.java (hunk 2 above)
Comment (ice1000) [reply pretty/src/main/java/org/aya/pretty/doc/Doc.java:252]: If you believe in some backend there will be special treatment for the symbol. For example, in LaTeX backend, `{` should be escaped as `\{`, and this is a special treatment, and hence you should use `symbol("{")`. In general, if you _feel_ like something is not some text, then probably symbol.

File: pretty/src/main/java/org/aya/pretty/doc/Doc.java (hunk 3 above)
Comment (ice1000) [reply pretty/src/main/java/org/aya/pretty/doc/Doc.java:534]: I see! Thanks for the explanation.

File: base/src/main/java/org/aya/distill/ConcreteDistiller.java [hunk 4]
```diff
@@ -153,7 +154,23 @@ yield visitCalls(false, fn, (nc, l) -> l.toDoc(options), outer,
         Doc.join(Doc.symbol("|"), idiom.barredApps().view()
           .map(app -> term(Outer.Free, app)))
       );
-      case Expr.Do aDo -> throw new UnsupportedOperationException("TODO");
+      case Expr.Do doExpr -> {
+        var doBlockDoc = doExpr.binds().map(bind -> {
+          if (bind.var() == LocalVar.IGNORED) {
+            return term(Outer.Free, bind.expr());
+          } else {
+            return visitDoBind(bind);
```
Comment (imkiva) [inline base/src/main/java/org/aya/distill/ConcreteDistiller.java:162]: We can omit these braces by `doExpr.binds().map(bind -> bind.var() == LocalVar.IGNORED ? term(Outer.Free, bind.expr()) : visitDoBind(bind));`

File: base/src/main/java/org/aya/distill/ConcreteDistiller.java [hunk 5]
```diff
@@ -312,6 +329,16 @@ private Stmt.Accessibility defaultAcc(@NotNull Decl.Personality personality) {
     };
   }
 
+  /**
+   * This function assumed that <code>doBind.var()</code> is not {@link org.aya.ref.LocalVar#IGNORED}
+   */
+  public @NotNull Doc visitDoBind(@NotNull Expr.DoBind doBind) {
```
Comment (imkiva) [inline base/src/main/java/org/aya/distill/ConcreteDistiller.java:335]: Put the `LocalVar.IGNORED` check here?

**Review by imkiva — COMMENTED** [review]
Looks good!

File: base/src/main/java/org/aya/distill/ConcreteDistiller.java [hunk 6]
```diff
@@ -155,21 +155,17 @@ yield visitCalls(false, fn, (nc, l) -> l.toDoc(options), outer,
...
 
         // Either not flat (single line) or full flat
-        yield Doc.flatAltBracedBlock(
-          Doc.commaList(doBlockDoc),
-          Doc.vcommaList(
-            doBlockDoc.map(x -> Doc.nest(2, x))
-          )
-        );
+        yield Doc.stickySep(      // doExpr is atom! It cannot be `do\n{ ... }`
+          Doc.symbol("do"),
```
Comment (imkiva) [inline base/src/main/java/org/aya/distill/ConcreteDistiller.java:162]: `Doc.styled(KEYWORD, "do")`

**Comment by ice1000** [comment]
> Where the test should be? I placed them in `ParseTest`, but its name is not good (We have to prevent the doExpr from desugaring)

Ιs there DistillerTest or PrettyTest?

**Comment by ice1000** [comment]
@HoshinoTented I merged your (another) PR and there are now conflicts. Can you resolve them?

**Comment by ice1000** [comment]
I merged your PR into another branch

**Comment by ice1000** [comment]
@HoshinoTented You can make it public

**Pull request closed without merging**